videostore/videostore.go

```diff
@@ -409,54 +441,90 @@ func (vs *videostore) processFrames(
-func (vs *videostore) deleter(ctx context.Context) {
-	ticker := time.NewTicker(deleterInterval * time.Minute)
+// fallbackDeleter is a go routine that cleans up old clips if storage is full.
+// It runs on interval and deletes the oldest clip(s) until the storage size is below the configured max.
+// The actual deletion logic should only execute when the indexer is failing to clean up on its own.
+// This is a worst case scenario prevention mechanism to ensure the disk doesn't fill up and lock up the system
+// if we end up in a state where the indexer isn't doing its job properly.
+func (vs *videostore) fallbackDeleter(ctx context.Context) {
+	// This is a best effort to avoid running the fallback deleter immediately on startup to let the indexer
+	// run its cleanup first.
+	select {
+	case <-time.After(fallbackDeleterStartupDelay * time.Minute):
+	case <-ctx.Done():
+		return
+	}
+
+	ticker := time.NewTicker(fallbackDeleterInterval * time.Minute)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			// Perform the deletion of the oldest clip
-			if err := cleanupStorage(vs.config.Storage.StoragePath, vs.config.Storage.SizeGB, vs.logger); err != nil {
-				vs.logger.Error("failed to clean up storage", err)
+			if err := vs.cleanupStorage(ctx); err != nil {
+				vs.logger.Error("fallback deleter failed to clean up storage: ", err)
 				continue
 			}
 		}
 	}
 }
 
-func cleanupStorage(storagePath string, maxStorageSizeGB int, logger logging.Logger) error {
-	maxStorageSize := int64(maxStorageSizeGB) * gigabyte
-	currStorageSize, err := getDirectorySize(storagePath)
+func (vs *videostore) cleanupStorage(ctx context.Context) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	// Apply multiplier to the configured max storage size to avoid clashing with the indexer
+	// so it only runs when the indexer is in a bad state.
+	fallbackStorageLimit := int64(float64(vs.config.Storage.SizeGB)*fallbackDeletionLimitMultiplier) * vsutils.Gigabyte
+	currStorageSize, err := vsutils.GetDirectorySize(vs.config.Storage.StoragePath)
 	if err != nil {
 		return err
 	}
-	if currStorageSize < maxStorageSize {
+	if currStorageSize < fallbackStorageLimit {
+		vs.logger.Debugf("storage size (%d bytes) is below fallback limit (%d bytes), no cleanup needed", currStorageSize, fallbackStorageLimit)
 		return nil
 	}
-	files, err := getSortedFiles(storagePath)
+
+	vs.logger.Error(
+		"storage usage is %.1fx the configured max storage size limit (%d bytes used > %d bytes limit), "+
+			"indexer is likely in an unhealthy state. Deleting oldest files manually to avoid system lockup",
+		fallbackDeletionLimitMultiplier,
+		currStorageSize,
+		fallbackStorageLimit,
+	)
+
+	files, err := vsutils.GetSortedFiles(vs.config.Storage.StoragePath)
 	if err != nil {
 		return err
 	}
 	for _, file := range files {
-		if currStorageSize < maxStorageSize {
+		if err := ctx.Err(); err != nil {
+			return err
```

Comment:
this isn't an error condition

Reply:
changing this return to nil makes the linter unhappy. I added a check in the caller of `cleanupStorage` instead